File: xdm-core/src/main/java/xdm/core/media/parser/dash/TemplateParser.java
package xdm.core.media.parser.dash;

import xdm.core.util.StringUtils;

import java.util.List;

public class TemplateParser {
  private TemplateParser() {}

  public static String parseTemplate(
      List<MatcherResult> matches,
      String templateUrl,
      long number,
      long time,
      String bandwidth,
      String representationId) {
    for (MatcherResult match : matches) {
      String variable = match.group(1);
      if (variable.startsWith("Number")) {
        if (variable.equals("Number")) {
          templateUrl = templateUrl.replace("$" + variable + "$", Long.toString(number));
        } else {
          String num = formatDigit(number, match, true);
          templateUrl = templateUrl.replace("$" + variable + "$", num);
        }
      } else if (variable.startsWith("Time")) {
        if (variable.equals("Time")) {
          templateUrl = templateUrl.replace("$" + variable + "$", Long.toString(time));
        } else {
          String num = formatDigit(time, match, false);
          templateUrl = templateUrl.replace("$" + variable + "$", num);
        }
      } else if (variable.equals("RepresentationID")) {
        templateUrl = templateUrl.replace("$" + variable + "$", representationId);
      } else if (variable.equals("Bandwidth")) {
        templateUrl = templateUrl.replace("$" + variable + "$", bandwidth);
      }
    }
    return templateUrl;
  }

  private static String formatDigit(long digit, MatcherResult match, boolean number) {
    String digitWidth = match.group(number ? "numdigits" : "timedigits");
    String dx = "d";
    if ((number && match.group("numdx") != null) || (!number && match.group("timedx") != null)) {
      dx = number ? match.group("numdx") : match.group("timedx");
    }
    int width = 0;
    if (!StringUtils.isNullOrEmpty(digitWidth)) {
      width = Integer.parseInt(digitWidth);
    }
    if (width > 0) {
      String fmt = String.format("%s%s%s", "%0", width, dx);
      return String.format(fmt, digit);
    } else {
      return Long.toString(digit);
    }
  }
}

File: xdm-core/src/main/java/xdm/core/downloaders/web/streaming/manifest/dash/TemplateParser.kt
package xdm.core.downloaders.web.streaming.manifest.dash


fun parseTemplate(
    matches: List<MatcherResult>,
    templateUrlIn: String,
    number: Long,
    time: Long,
    bandwidth: String,
    representationId: String
): String {
    var templateUrl = templateUrlIn
    for (match in matches) {
        val variable = match.group(1)
        if (variable?.startsWith("Number") == true) {
            if (variable == "Number") {
                templateUrl = templateUrl.replace("$$variable$", number.toString())
            } else {
                val num = formatDigit(number, match, true)
                templateUrl = templateUrl.replace("$$variable$", num)
            }
        } else if (variable?.startsWith("Time") == true) {
            if (variable == "Time") {
                templateUrl = templateUrl.replace("$$variable$", time.toString())
            } else {
                val num = formatDigit(time, match, false)
                templateUrl = templateUrl.replace("$$variable$", num)
            }
        } else if (variable == "RepresentationID") {
            templateUrl = templateUrl.replace("$$variable$", representationId)
        } else if (variable == "Bandwidth") {
            templateUrl = templateUrl.replace("$$variable$", bandwidth)
        }
    }
    return templateUrl
}

private fun formatDigit(digit: Long, match: MatcherResult, number: Boolean): String {
    val digitWidth = match.group(if (number) "numdigits" else "timedigits")
    var dx: String? = "d"
    if ((number && match.group("numdx") != null) || (!number && match.group("timedx") != null)) {
        dx = if (number) match.group("numdx") else match.group("timedx")
    }
    var width = 0
    if (!digitWidth.isNullOrEmpty()) {
        width = digitWidth.toInt()
    }
    return if (width > 0) {
        val fmt = String.format("%s%s%s", "%0", width, dx)
        String.format(fmt, digit)
    } else {
        digit.toString()
    }
}
